File: bot/evolution.py
import logging
import requests
from config import EVOLUTION_URL, EVOLUTION_API_KEY

log = logging.getLogger('evolution')

_HEADERS = {
    'apikey': EVOLUTION_API_KEY or '',
    'Content-Type': 'application/json'
}

# In-memory cache: LID JID -> resolved phone number
_lid_cache = {}


def resolve_lid_to_phone(instance_name, lid_jid):
    """Resolve a LID JID to a real phone number using Evolution contacts API.

    WhatsApp uses LID (Linked ID) format for some contacts. The Evolution API
    cannot send messages to LID addresses directly, so we resolve them by
    matching profilePicUrl between the LID contact and @s.whatsapp.net contacts.
    """
    if lid_jid in _lid_cache:
        return _lid_cache[lid_jid]

    try:
        # Get all contacts for this instance
        r = requests.post(
            f'{EVOLUTION_URL}/chat/findContacts/{instance_name}',
            headers=_HEADERS,
            json={},
            timeout=10
        )
        if r.status_code != 200:
            log.warning(f'findContacts falhou ({r.status_code})')
            return None

        contacts = r.json()
        if not isinstance(contacts, list):
            return None

        # Find the LID contact
        lid_contact = None
        for c in contacts:
            if c.get('remoteJid') == lid_jid:
                lid_contact = c
                break

        if not lid_contact:
            return None

        pic_url = lid_contact.get('profilePicUrl')
        push_name = lid_contact.get('pushName', '')

        # Strategy 1: Match by profilePicUrl (most reliable)
        if pic_url:
            for c in contacts:
                rjid = c.get('remoteJid', '')
                if rjid.endswith('@s.whatsapp.net') and c.get('profilePicUrl') == pic_url:
                    phone = rjid.split('@')[0]
                    _lid_cache[lid_jid] = phone
                    log.info(f'LID resolvido via profilePic: {lid_jid} -> {phone}')
                    return phone

        # Strategy 2: Match by pushName (less reliable, only if unique match)
        if push_name:
            candidates = [
                c for c in contacts
                if c.get('remoteJid', '').endswith('@s.whatsapp.net')
                and c.get('pushName') == push_name
            ]
            if len(candidates) == 1:
                phone = candidates[0]['remoteJid'].split('@')[0]
                _lid_cache[lid_jid] = phone
                log.info(f'LID resolvido via pushName: {lid_jid} -> {phone}')
                return phone

        log.warning(f'Nao foi possivel resolver LID: {lid_jid}')
        return None
    except Exception as e:
        log.error(f'Erro ao resolver LID: {e}')
        return None


def set_typing(instance_name, phone, typing=True):
    try:
        requests.post(
            f'{EVOLUTION_URL}/chat/updatePresence/{instance_name}',
            headers=_HEADERS,
            json={'number': phone, 'presence': 'composing' if typing else 'paused'},
            timeout=3
        )
    except Exception:
        pass


def send_message(instance_name, phone, text):
    try:
        r = requests.post(
            f'{EVOLUTION_URL}/message/sendText/{instance_name}',
            headers=_HEADERS,
            json={'number': phone, 'text': text},
            timeout=10
        )
        if r.status_code in (200, 201):
            return True
        log.warning(f'Envio falhou ({r.status_code}): {r.text[:200]}')
        # fallback: formato alternativo (textMessage wrapper)
        r = requests.post(
            f'{EVOLUTION_URL}/message/sendText/{instance_name}',
            headers=_HEADERS,
            json={'number': phone, 'textMessage': {'text': text}},
            timeout=10
        )
        if r.status_code in (200, 201):
            return True
        log.error(f'Envio fallback falhou ({r.status_code}): {r.text[:200]}')
        return False
    except Exception as e:
        log.error(f'Erro envio: {e}')
        return False


def create_instance(instance_name):
    r = requests.post(
        f'{EVOLUTION_URL}/instance/create',
        headers=_HEADERS,
        json={
            'instanceName': instance_name,
            'qrcode': True,
            'integration': 'WHATSAPP-BAILEYS'
        },
        timeout=15
    )
    return r.json() if r.status_code in (200, 201) else {'error': r.text}


def get_connection_state(instance_name):
    try:
        r = requests.get(
            f'{EVOLUTION_URL}/instance/connectionState/{instance_name}',
            headers=_HEADERS,
            timeout=5
        )
        return r.json().get('instance', {}).get('state', 'unknown')
    except Exception:
        return 'error'


def get_qr_code(instance_name):
    try:
        r = requests.get(
            f'{EVOLUTION_URL}/instance/connect/{instance_name}',
            headers=_HEADERS,
            timeout=10
        )
        return r.json() if r.status_code == 200 else {'error': r.text}
    except Exception as e:
        return {'error': str(e)}


def fetch_all_instances():
    try:
        r = requests.get(
            f'{EVOLUTION_URL}/instance/fetchInstances',
            headers=_HEADERS,
            timeout=10
        )
        return r.json() if r.status_code == 200 else []
    except Exception:
        return []


def delete_instance(instance_name):
    try:
        r = requests.delete(
            f'{EVOLUTION_URL}/instance/delete/{instance_name}',
            headers=_HEADERS,
            timeout=10
        )
        return r.json()
    except Exception as e:
        return {'error': str(e)}


def logout_instance(instance_name):
    try:
        r = requests.delete(
            f'{EVOLUTION_URL}/instance/logout/{instance_name}',
            headers=_HEADERS,
            timeout=10
        )
        return r.json()
    except Exception as e:
        return {'error': str(e)}


def set_webhook(instance_name, webhook_url):
    try:
        r = requests.post(
            f'{EVOLUTION_URL}/webhook/set/{instance_name}',
            headers=_HEADERS,
            json={
                'webhook': {
                    'enabled': True,
                    'url': webhook_url,
                    'webhookByEvents': True,
                    'events': ['MESSAGES_UPSERT']
                }
            },
            timeout=10
        )
        return r.status_code == 200
    except Exception:
        return False

File: agent-node/src/integrations/evolution.js
const axios = require('axios');
const config = require('../config');
const logger = require('../utils/logger');
const lidResolver = require('../services/lid-resolver');

class EvolutionClient {
  constructor() {
    this.baseUrl = config.evolution.url;
    this.apiKey = config.evolution.apiKey;
  }

  getHeaders() {
    return {
      'Content-Type': 'application/json',
      'apikey': this.apiKey,
    };
  }

  /**
   * Send a text message with LID resolution
   */
  async sendMessage(instance, remoteJid, text) {
    try {
      // Resolve LID to phone number if needed
      const destinationJid = await lidResolver.getDestinationJid(instance, remoteJid);

      logger.info(`[SEND] Instance: ${instance}, To: ${destinationJid}`);

      const url = `${this.baseUrl}/message/sendText/${instance}`;

      const response = await axios.post(url, {
        number: destinationJid,
        text: text,
      }, {
        headers: this.getHeaders(),
        timeout: 15000,
      });

      if (response.status === 200 || response.status === 201) {
        logger.info(`[OK] Message sent to ${destinationJid}`);
        return response.data;
      }

      // Fallback: try with textMessage wrapper
      logger.warn(`First attempt failed, trying textMessage wrapper...`);
      const fallbackResponse = await axios.post(url, {
        number: destinationJid,
        textMessage: { text: text },
      }, {
        headers: this.getHeaders(),
        timeout: 15000,
      });

      if (fallbackResponse.status === 200 || fallbackResponse.status === 201) {
        logger.info(`[OK] Message sent via fallback to ${destinationJid}`);
        return fallbackResponse.data;
      }

      throw new Error(`Unexpected status: ${fallbackResponse.status}`);

    } catch (error) {
      const errorMsg = error.response?.data?.message || error.response?.data || error.message;
      logger.error(`[FAIL] Send message error: ${JSON.stringify(errorMsg)}`);
      throw new Error(`Failed to send message: ${JSON.stringify(errorMsg)}`);
    }
  }

  /**
   * Send an audio message (voice note)
   */
  async sendAudio(instance, remoteJid, base64Audio) {
    try {
      const destinationJid = await lidResolver.getDestinationJid(instance, remoteJid);

      logger.info(`[SEND AUDIO] Instance: ${instance}, To: ${destinationJid}`);

      const url = `${this.baseUrl}/message/sendWhatsAppAudio/${instance}`;

      const response = await axios.post(url, {
        number: destinationJid,
        audio: base64Audio,
      }, {
        headers: this.getHeaders(),
        timeout: 30000,
      });

      if (response.status === 200 || response.status === 201) {
        logger.info(`[OK] Audio sent to ${destinationJid}`);
        return response.data;
      }

      throw new Error(`Unexpected status: ${response.status}`);

    } catch (error) {
      const errorMsg = error.response?.data?.message || error.response?.data || error.message;
      logger.error(`[FAIL] Send audio error: ${JSON.stringify(errorMsg)}`);
      throw new Error(`Failed to send audio: ${JSON.stringify(errorMsg)}`);
    }
  }

  /**
   * Send media (image, video, document)
   */
  async sendMedia(instance, remoteJid, mediaUrl, caption = '', mediaType = 'image') {
    try {
      const destinationJid = await lidResolver.getDestinationJid(instance, remoteJid);

      logger.info(`[SEND MEDIA] Instance: ${instance}, To: ${destinationJid}, Type: ${mediaType}`);

      const url = `${this.baseUrl}/message/sendMedia/${instance}`;

      const response = await axios.post(url, {
        number: destinationJid,
        mediatype: mediaType,
        media: mediaUrl,
        caption: caption,
      }, {
        headers: this.getHeaders(),
        timeout: 30000,
      });

      if (response.status === 200 || response.status === 201) {
        logger.info(`[OK] Media sent to ${destinationJid}`);
        return response.data;
      }

      throw new Error(`Unexpected status: ${response.status}`);

    } catch (error) {
      const errorMsg = error.response?.data?.message || error.response?.data || error.message;
      logger.error(`[FAIL] Send media error: ${JSON.stringify(errorMsg)}`);
      throw new Error(`Failed to send media: ${JSON.stringify(errorMsg)}`);
    }
  }

  /**
   * Set typing indicator
   */
  async setTyping(instance, remoteJid, isTyping = true) {
    try {
      const destinationJid = await lidResolver.getDestinationJid(instance, remoteJid);

      await axios.post(
        `${this.baseUrl}/chat/updatePresence/${instance}`,
        {
          number: destinationJid,
          presence: isTyping ? 'composing' : 'paused',
        },
        {
          headers: this.getHeaders(),
          timeout: 5000,
        }
      );

      logger.debug(`Typing ${isTyping ? 'started' : 'stopped'} for ${destinationJid}`);
    } catch (error) {
      // Non-critical, just log
      logger.debug(`Typing indicator failed: ${error.message}`);
    }
  }

  /**
   * Download media as base64
   */
  async getBase64Media(instance, messageKey) {
    try {
      const response = await axios.post(
        `${this.baseUrl}/chat/getBase64FromMediaMessage/${instance}`,
        { message: { key: messageKey } },
        {
          headers: this.getHeaders(),
          timeout: 30000,
        }
      );

      if (response.status === 200 || response.status === 201) {
        logger.info(`[OK] Media downloaded`);
        return response.data.base64 || '';
      }

      return '';
    } catch (error) {
      logger.error(`[FAIL] Media download: ${error.message}`);
      return '';
    }
  }

  /**
   * Get instance connection status
   */
  async getInstanceStatus(instance) {
    try {
      const response = await axios.get(
        `${this.baseUrl}/instance/connectionState/${instance}`,
        { headers: this.getHeaders(), timeout: 5000 }
      );

      const state = response.data?.instance?.state || 'unknown';
      logger.debug(`Instance ${instance} state: ${state}`);
      return state;

    } catch (error) {
      logger.error(`Instance status error: ${error.message}`);
      return 'error';
    }
  }

  /**
   * Fetch all contacts for LID resolution
   */
  async fetchContacts(instance) {
    try {
      const response = await axios.post(
        `${this.baseUrl}/chat/findContacts/${instance}`,
        {},
        { headers: this.getHeaders(), timeout: 10000 }
      );

      return response.data || [];
    } catch (error) {
      logger.error(`Fetch contacts error: ${error.message}`);
      return [];
    }
  }
}

module.exports = new EvolutionClient();
